docs/internal/versioning.md

```diff
@@ -0,0 +1,244 @@
+Versioning in Verona
+====================
+
+**WARNING** This is an under-discussion proposal, not a final design for the language.
+
+Verona, as an infrastructure language, aims to support large-scale software engineering, where programs may be built from different components with different release schedules and that are maintained by different teams.
+A Verona programmer should have the expectation that updating to a new version of a package will not break their code.
+A Verona programmer publishing a library should have tooling that improves their ability to make this guarantee for their downstream consumers.
+
+The key goals for versioning are:
+
+ - A programmer should be able to ship breaking changes to an API with backwards-compatibility shims and graceful deprecation.
+ - Updating to a new version of a dependency should not cause code to fail to compile unless your code depends on a version of the interface that has been explicitly dropped in the new version.
+ - Developers should be given tooling at warns them of API breakages and allows them to provide compatibility interfaces.
+ - If a library ships a new version of an interface, downstream consumers should be able to gradually upgrade their codebase in the common case (unless the new API provides fundamentally different abstractions).
+
+Visibility is a software-engineering feature
+--------------------------------------------
+
+Many object-oriented languages provide some forms of restrictions on when the fields or methods of an object can be accessed.
+The Java security model depends on these.
+In Verona, visibility is regarded as purely a software-engineering tool for defining API-stability guarantees.
+As such, Verona provides only two levels of visibility:
+
+ - Public: may be accessed by anything
+ - Package-private: may be accessed only by code written in the same package
```

Comment:
We don't think it's useful.  There are two reasons for type-private:

 - You want to build some stronger security properties on top and do very fine-grained isolation that depends on this.
 - You want to guarantee API stability by preventing anything from depending on the implementation details.

The first of these was embraced by Java and leads to the kind of 'defence in breadth' design that the JVM ended up with, where a bug in any part of the implementation can violate the security properties.  We don't want that.

For the second part, type-level private always ends up needing escape hatches like `friend` declarations in C++ or undocumented (or documented 'do-not-use') methods in Smalltalk-family languages because it's actually fine for code that is always shipped along with a class to depend on implementation details of that class.  You end up with cognitive load (remember trying to explain what `public`, `private`, and `package` mean and when they should be used to undergrads?) for a thing that provides no real value: it doesn't add security and it doesn't make software engineering easier.